### ISSUE TITLE
PP_OCRv2识别模型推理时为什么一定要做resize？

### ISSUE THREAD

**Issue by Smile0524** [PaddlePaddle/PaddleOCR#11788]
请提供下述完整信息以便快速定位问题/Please provide the following information to quickly locate the problem

- 系统环境/System Environment：
- 版本号/Version：Paddle：  PaddleOCR： 问题相关组件/Related components：
- 运行指令/Command Code：
- 完整报错/Complete Error Message：

请尽量不要包含图片在问题中/Please try to not include the image in the issue.

在识别模型的推理代码中，加入了resize部分，导致和训练时resize的尺寸不一样，最后导致训练与推理的结果不一致，这部分怎么解决？
    def resize_norm_img(self, img, max_wh_ratio):
        imgC, imgH, imgW = self.rec_image_shape
        if self.rec_algorithm == 'NRTR':
            img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
            # return padding_im
            image_pil = Image.fromarray(np.uint8(img))
            img = image_pil.resize([100, 32], Image.ANTIALIAS)
            img = np.array(img)
            norm_img = np.expand_dims(img, -1)
            norm_img = norm_img.transpose((2, 0, 1))
            return norm_img.astype(np.float32) / 128. - 1.

        assert imgC == img.shape[2]
#### 这一句：
        imgW = int((32 * max_wh_ratio))
        if self.use_onnx:
            imgW = 100
        h, w = img.shape[:2]
        ratio = w / float(h)
        if math.ceil(imgH * ratio) > imgW:
            resized_w = imgW
        else:
            resized_w = int(math.ceil(imgH * ratio))
        resized_image = cv2.resize(img, (resized_w, imgH))
        resized_image = resized_image.astype('float32')
        resized_image = resized_image.transpose((2, 0, 1)) / 255
        resized_image -= 0.5
        resized_image /= 0.5
        padding_im = np.zeros((imgC, imgH, imgW), dtype=np.float32)
        padding_im[:, :, 0:resized_w] = resized_image
        return padding_im

**Comment by Smile0524** [comment]
补充：因为涉及到导包paddleocr，直接注释掉这一行需要改动源码，比较麻烦，请问是否有其他解决办法？

**Comment by changdazhou** [comment]
辛苦提供一下你的训练和推理的命令哈

**Comment by Smile0524** [comment]
训练：python3 tools/train.py -c configs/rec/ch_PP-OCRv2/ch_PP-OCRv2_rec.yml | tee train_rec.log

**Comment by Smile0524** [comment]
训练：python3 tools/train.py -c configs/rec/ch_PP-OCRv2/ch_PP-OCRv2_rec.yml | tee train_rec.log
推理：采用的PaddleOCR(***).ocr()

**Comment by Smile0524** [comment]
训练config:
Global:
  debug: false
  use_gpu: false
  epoch_num: 500
  log_smooth_window: 20
  print_batch_step: 10
  save_model_dir: ../../out_model/PP-OCRv2/
  save_epoch_step: 50
  eval_batch_step: [0, 200]
  cal_metric_during_train: true
  pretrained_model:
  checkpoints:
  save_inference_dir:
  use_visualdl: false
  infer_img: doc/imgs_words/ch/word_1.jpg
  character_dict_path: ppocr/utils/only_char_digit_keys.txt
  max_text_length: 25
  infer_mode: false
  use_space_char: true
  distributed: true
  save_res_path: ../../out_model/PP-OCRv2.txt


Optimizer:
  name: Adam
  beta1: 0.9
  beta2: 0.999
  lr:
    name: Piecewise
    decay_epochs : [700]
    values : [0.001, 0.0001]
    warmup_epoch: 5
  regularizer:
    name: L2
    factor: 2.0e-05


Architecture:
  model_type: rec
  algorithm: CRNN
  Transform:
  Backbone:
    name: MobileNetV1Enhance
    scale: 0.5
  Neck:
    name: SequenceEncoder
    encoder_type: rnn
    hidden_size: 64
  Head:
    name: CTCHead
    mid_channels: 96
    fc_decay: 0.00002

Loss:
  name: CTCLoss

PostProcess:
  name: CTCLabelDecode

Metric:
  name: RecMetric
  main_indicator: acc

Train:
  dataset:
    name: SimpleDataSet
    data_dir: ../../data/handwrite_rec/rec/train
    label_file_list:
    - ../../data/handwrite_rec/rec/train.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - RecAug:
    - CTCLabelEncode:
    - RecResizeImg:
        image_shape: [3, 32, 320]
    - KeepKeys:
        keep_keys:
        - image
        - label
        - length
  loader:
    shuffle: true
    batch_size_per_card: 256
    drop_last: true
    num_workers: 8
Eval:
  dataset:
    name: SimpleDataSet
    data_dir: ../../data/handwrite_rec/rec/val
    label_file_list:
    - ../../data/handwrite_rec/rec/val.txt
    transforms:
    - DecodeImage:
        img_mode: BGR
        channel_first: false
    - CTCLabelEncode:
    - RecResizeImg:
        image_shape: [3, 32, 320]
    - KeepKeys:
        keep_keys:
        - image
        - label
        - length
  loader:
    shuffle: false
    drop_last: false
    batch_size_per_card: 1
    num_workers: 8

**Comment by changdazhou** [comment]
def resize_norm_img(self, img, max_wh_ratio):
imgC, imgH, imgW = self.rec_image_shape
if self.rec_algorithm == 'NRTR':
img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
# return padding_im
image_pil = Image.fromarray(np.uint8(img))
img = image_pil.resize([100, 32], Image.ANTIALIAS)
img = np.array(img)
norm_img = np.expand_dims(img, -1)
norm_img = norm_img.transpose((2, 0, 1))
return norm_img.astype(np.float32) / 128. - 1.

**Comment by changdazhou** [comment]
def resize_norm_img(self, img, max_wh_ratio):
imgC, imgH, imgW = self.rec_image_shape
if self.rec_algorithm == 'NRTR':
img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
# return padding_im
image_pil = Image.fromarray(np.uint8(img))
img = image_pil.resize([100, 32], Image.ANTIALIAS)
img = np.array(img)
norm_img = np.expand_dims(img, -1)
norm_img = norm_img.transpose((2, 0, 1))
return norm_img.astype(np.float32) / 128. - 1.
这段代码是哪里的呢，一般不建议推理和训练图片尺寸不一致哈

**Comment by Smile0524** [comment]
在paddleocr/tools/infer/predict_rec.py里，这个代码里会自动调尺寸，不是我能控制的。训练的时候都是在config文件里指定宽度训练的，不知道为啥代码里推理的时候，会调整宽度，这个导致最后训练、推理结果不一致。

**Comment by Smile0524** [comment]
可能没明白我的意思，推理用的代码是paddleocr源码里写的，这个不是我写的。

**Comment by changdazhou** [comment]
看了一下，你这个使用的是2.5分支的paddleocr，建议下载最新源码使用最新版本的PaddleOCR，应该是因为之前的代码不适配导致的
